src/BenchmarkDotNet.Disassembler.x64/Program.cs

```diff
@@ -274,177 +237,146 @@ private static string GetSmartPrefix(string sourceLine, int length)
             return new string(prefix);
         }
 
-        static IEnumerable<Asm> GetAsm(ILToNativeMap map, State state, int depth, ClrMethod currentMethod)
+        private static IEnumerable<Asm> GetAsm(ILToNativeMap map, State state, int depth, ClrMethod currentMethod)
         {
-            var disasmBuffer = new StringBuilder(512);
-            ulong disasmAddress = map.StartAddress;
-            while (true)
-            {
-                int hr = state.DebugControl.Disassemble(disasmAddress, 0,
-                    disasmBuffer, disasmBuffer.Capacity, out uint disasmSize,
-                    out ulong endOffset);
-                if (hr != 0)
-                    break;
+            int length = (int)(map.EndAddress - map.StartAddress);
+            byte[] buffer = new byte[length];
+
+            if (!state.Runtime.DataTarget.ReadProcessMemory(map.StartAddress, buffer, buffer.Length, out int bytesRead))
```

Comment:
I think this can/should be protected with some extra checking, on Linux, since many users will probably hit this head-on.

Perhaps the correct place is **not here** but rather when the Diagnoser is condigured at the top level when the run is about to begin...?

I'm specifically referring to `ptrace_scope` on linux. If I'm not mistaken, that kernel flag controls whether `ReadProcessMemory` is going to actually work (Again ,95% sure this is the one, I'm doing this from memory). I personally just set it to `0` in my `sysctl.conf`, which means I can debug my own process (same uid)...

Here is some rough documentation copy-pasted from [here](https://linux-audit.com/protect-ptrace-processes-kernel-yama-ptrace_scope/):

```bash
$ cat /proc/sys/kernel/yama/ptrace_scope
1
```

For this particular key there are four valid options: 0-3
* kernel.yama.ptrace_scope = 0: all processes can be debugged, as long as they have same uid. This is the classical way of how ptracing worked.
* kernel.yama.ptrace_scope = 1: only a parent process can be debugged.
* kernel.yama.ptrace_scope = 2: Only admin can use ptrace, as it required CAP_SYS_PTRACE capability.
* kernel.yama.ptrace_scope = 3: No processes may be traced with ptrace. Once set, a reboot is needed to enable ptracing again.

I *think* it makes sense to check that `uid == 0` or that ptrace_scope is `0`.

Otherwise, you will have to make sure that ptrace_scope is `1` AND that the code you WANT to disassemble lanches the disassembler (I have no idea if that is the current case, just a general suggestion...)...

Of course values: `2`,`3` can also be handled appropriately...

Does this sound good?

Reply:
@damageboy thanks for pointing this out! If you have not told me about this, I would most probably learn about this with the first user bug report.

I've added proper handling of `ptrace_scope` value. When it's 3 it's an error, when it's 2 it's a warning. With messages ofc.

BTW the LinuxDisassembler works differently than Windows one. We don't spawn a dedicated disassembler process because the parent and benchmark process have always the same architecture on Linux (on Windows we have x86 and x64, on Linux .NET Core supports only x64)